### PR TITLE
Fix DLNA issues in 10.6

### LINKED ISSUE #1634: DLNA not working in 10.6
From https://www.reddit.com/r/airsonic/comments/g2zdx0/dlna_and_106/:

> I have been testing 10.6, but no matter how I fiddle the DLNA server remains invisible. I run Airsonic in a war container in windows (sush), and normally use bubbleupnp too send the tracks to my end point a NAD C338.
>
> I tried both running 10.6 in the old 10.51 folder,letting it use the old database and also a completely fresh install. Bubble couldnt see either install. Replacing the war with the old one and touching nothing else, the airsonic server was back.
>
> Maybe I am missing something obvious?. For now i'm back too 10.51 which works fine in my setup.

### PR DESCRIPTION
**Pull request by fxthomas**
Fixes #1634.

### CODE REVIEW

**Comment by fxthomas** [comment]
I don't use DLNA, so this was a bit of a wild shot on a technology I don't understand well yet.

From what I understand : Cling (the library we use for UPnP/DLNA) does some weird hacks in `HttpUrlConnection` which don't play well with servlet containers like Tomcat (whether embedded or external). A workaround is available in Cling 2.0 in the form of helper classes (which were used in Airsonic before d626463), but is removed in Cling 2.1.

Relevant issues: #1505, #1381, #978, #975.

**Comment by tesshucom** [comment]
As for the core of the problem, it's accurate you presented.
Previously I made a similar fix in the clone.
(There are more UPnP / DLNA users in Japan than in other countries. Therefore a careful approach was taken.)

https://github.com/tesshucom/jpsonic/commit/24d54fd36d5f89e93cc0ade31d14b9959da7cb73

Ephemeral port is the default in Cling.
(To be honest, the concrete value is 0, but it makes sense that Airsonic has no default value.)
Subsonic, Libresonic and Airsonic have been like that until now.

As you know, this has changed in 10.6.0.
I did not follow this and fixed it to allow port changes without breaking the legacy default design.

Demonstrating the cases in which this has a net effect is time consuming, but still a lower risk to existing users.
Because it's been working that way until now.
I've taken the definitive way to avoid having to verify on Windows, which has a unique auto-port configuration feature.
(Since the settings differ depending on the version, you have to verify it individually each time trouble occurs!)

I don't care how Airsonic handles the UPnP default port.
However, it has not been verified whether the default port shows a different value from the legacy server on a particular OS.

If you don't want to have any new validations or issues, then a switch like the one above.

File: airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public ApacheUpnpServiceConfiguration(int streamListenPort) {
+        super(streamListenPort);
+    }
+
+    @Override
+    public StreamClient createStreamClient() {
+        return new StreamClientImpl(new StreamClientConfigurationImpl(Executors.newCachedThreadPool()));
+    }
```
Comment (randomnicode) [inline airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java:53]: This is again more or less the same as the parent. The only difference is that the parent provides its own custom executor and we provide a generic one.

File: airsonic-main/src/main/java/org/airsonic/player/service/UPnPService.java [hunk 2]
```diff
@@ -122,7 +122,7 @@ private void startService() {
     }
 
     private synchronized void createService() {
-        upnpService = new UpnpServiceImpl(new DefaultUpnpServiceConfiguration(SettingsService.getDefaultUPnPPort()));
+        upnpService = new UpnpServiceImpl(new ApacheUpnpServiceConfiguration(SettingsService.getDefaultUPnPPort()));
```
Comment (randomnicode) [inline airsonic-main/src/main/java/org/airsonic/player/service/UPnPService.java:125]: I do not see much of a difference between this custom class and the parent class except for an request executor. I'm not convinced we know the issue.

**Comment by randomnicode** [comment]
Note, I'm not denying there isn't an issue, or even that this doesn't fix it, just that it seems to me that the underlying root cause isn't understood at all.

In my mind, you may as well just fix it with a much shorter version of the PR. An almost one liner by overriding the createDefaultExecutorService() anonymously instead of creating a whole new separate class file:

```
upnpService = new UpnpServiceImpl(new 
        DefaultUpnpServiceConfiguration(settingsService.getUPnpPort()) {
            protected ExecutorService createDefaultExecutorService() {
                return Executors.newCachedThreadPool();
            }
        });
```

This seems to be the only change you're effecting by creating a separate class and overriding a bunch of methods.

File: airsonic-main/src/main/java/org/airsonic/player/service/UPnPService.java (hunk 2 above)
Comment (randomnicode) [reply airsonic-main/src/main/java/org/airsonic/player/service/UPnPService.java:125]: Scratch that, I think I see what you're saying: the packages are different for the similarly named the classes

File: airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java (hunk 1 above)
Comment (randomnicode) [reply airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java:53]: Nevermind, the packages are different

File: airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public StreamServer createStreamServer(NetworkAddressFactory networkAddressFactory) {
+        return new StreamServerImpl(new StreamServerConfigurationImpl(networkAddressFactory.getStreamListenPort()));
+    }
```
[the remark this answers is not in the document]
Comment (randomnicode) [reply airsonic-main/src/main/java/org/airsonic/player/service/upnp/ApacheUpnpServiceConfiguration.java:58]: Different packages

**Comment by randomnicode** [comment]
Yeah, my apologies, I stand completely corrected. I see the issue you're pointing out now.

In that case, this looks perfectly fine.

One minor nit would be to switch the cached threadpool to whatever the class's parent (Cling's custom threadpool) is using. I haven't looked in any detail, but no reason to deviate from the parent? (famous last words...)

On a side note, as you pointed, Cling is now unmaintained, and perhaps we should switch or stop providing DLNA (which is kinda getting old). A more recent fork of cling seems to be here: https://github.com/jupnp/jupnp

**Comment by tesshucom** [comment]
We are very aware of that.
What is happening now is a reinstatement to deal with the feature degradation Instantly.
Also, it is an issue pointed out by the user .
It makes sense to eliminate noise and treat it as an incident response.

Other improvements can be treated as separate commits.
Blurring the problem of degradation and how to deal with it is not appropriate.

The reason why long-standing Airsonic maintainers recommend splitting (although this can be overkill) is that they are wary of such incidents.

**Comment by fxthomas** [comment]
@tesshucom I think I will leave the default port for now (although you're right, it's maybe a commonly-used port, hopefully it won't cause any issues). Thanks for looking at this, merging!

**Comment by fxthomas** [comment]
@randomnicode For now Cling doesn't cause issues with Tomcat, but I would be very wary of Tomcat upgrades while we stick to this.

**Pull request merged**